### PR TITLE
Implement basic input layer

### LINKED ISSUE #35: Implement basic input layer

### PR DESCRIPTION
**Pull request by AndreySorokin7**
Closes #35

### CODE REVIEW

File: include/CMakeLists.txt [hunk 1]
```diff
@@ -8,3 +8,18 @@ file(GLOB_RECURSE perf_headers perf/*.h perf/*.hpp)
...
+file(MAKE_DIRECTORY "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build")
+
+execute_process(
+    COMMAND ${CMAKE_COMMAND} -S "${CMAKE_SOURCE_DIR}/3rdparty/opencv" -B "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build" -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+    WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build"
+)
+execute_process(
+    COMMAND ${CMAKE_COMMAND} --build "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build" --config "${CMAKE_BUILD_TYPE}"
+    WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build"
+)
```
Comment (aobolensk) [inline include/CMakeLists.txt:21]: Why do you build OpenCV one more time?

Comment (AndreySorokin7) [reply include/CMakeLists.txt:21]: Because the person who worked before connected reading only for the "read" folder and I did not see OpenCV in the folder from which I worked. So I rewrote include for everything. And please see what's wrong with the bot. Maybe I'm doing something wrong and I don't understand, but it seems like I took the changes from the main, but the situation is the same.

Comment (aobolensk) [reply include/CMakeLists.txt:21]: Need to move OpenCV build to 3rdparty. Cause here we are duplicating OpenCV build. As for the bot, consider this as a non-blocking issue, if we don't find the root cause until all comments resolved and 3 approvals collected, we'll merge it in any case

Comment (AndreySorokin7) [reply include/CMakeLists.txt]: fixed

File: include/layers/InputLayer.hpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+ public:
+  InputLayer(const int& Cpic) : Layer<ValueType>() { Count_pic_ = Cpic; }
+  std::vector<ValueType> run(const std::vector<ValueType>& input) const {
+    std::vector<ValueType> a = input;
+    return a;
+  }
+  Tensor run(const std::vector<std::string>& path) const {
+    std::vector<int> res(Count_pic_ * 227 * 227 * 3);
+    for (int num = 0; num < Count_pic_; num++) {
```
Comment (NeiroYT) [inline include/layers/InputLayer.hpp:25]: Check if path.size() == Count_pic_ or don't use Count_pic_

File: include/layers/InputLayer.hpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      int c = 0;
+      for (int i = 0; i < 227; ++i) {
+        for (int j = 0; j < 227; ++j) {
+          res[c] = static_cast<int>(channels[2].at<uchar>(i, j));
+          c++;
+          res[c] = static_cast<int>(channels[1].at<uchar>(i, j));
+          c++;
+          res[c] = static_cast<int>(channels[0].at<uchar>(i, j));
+          c++;
+        }
```
Comment (NeiroYT) [inline include/layers/InputLayer.hpp:43]: you can define Tensor res({Count_pic_, 227, 227, 3}) and do all these operation with set() function

File: include/layers/InputLayer.hpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+#include <cstdlib>
+#include <iostream>
+#include <opencv2/opencv.hpp>
+#include <stdexcept>
+#include <string>
+
+#include "layers/Layer.hpp"
+
+template <typename ValueType>
+class InputLayer : public Layer<ValueType> {
```
Comment (NeiroYT) [inline include/layers/InputLayer.hpp:13]: Should be not template and not depend from Layerimpl

**Review by NeiroYT — CHANGES_REQUESTED** [review]

File: 3rdparty/CMakeLists.txt [hunk 5]
```diff
@@ -1 +1,12 @@
...
+file(MAKE_DIRECTORY "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build")
+
+execute_process(
+    COMMAND ${CMAKE_COMMAND} -S "${CMAKE_SOURCE_DIR}/3rdparty/opencv" -B "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build" -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+    WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build"
+)
+execute_process(
+    COMMAND ${CMAKE_COMMAND} --build "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build" --config "${CMAKE_BUILD_TYPE}"
+    WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build"
+)
```
Comment (aobolensk) [inline 3rdparty/CMakeLists.txt:12]: Move this change to the separate PR

File: include/CMakeLists.txt [hunk 6]
```diff
@@ -9,3 +9,7 @@ set(PERF_HEADERS "${perf_headers}" PARENT_SCOPE)
 
 add_subdirectory(alexnet)
 add_subdirectory(read)
+
+find_package( OpenCV REQUIRED PATHS "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build" )
+include_directories( ${OpenCV_INCLUDE_DIRS} )
+target_link_libraries( ReadLib ${OpenCV_LIBS} )
```
Comment (aobolensk) [inline include/CMakeLists.txt:15]: Same, to the separate change

File: app/CMakeLists.txt [hunk 7]
```diff
@@ -20,6 +20,7 @@ file(DOWNLOAD
     STATUS status_code
     LOG log_file
 )
+
```
Comment (aobolensk) [inline app/CMakeLists.txt:23]:
```suggestion
```

File: include/layers/InputLayer.hpp [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+#include <stdexcept>
+#include <string>
+
+#include "layers/Layer.hpp"
+
+class InputLayer : public Layer {
+ public:
+  InputLayer() = default;
+  static void run(const std::vector<std::string>& path, Tensor& output) {
+    std::vector<int> res(path.size() * 227 * 227 * 3);
```
Comment (aobolensk) [inline include/layers/InputLayer.hpp:16]: Shape should not be hardcoded

File: include/layers/InputLayer.hpp [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+#include <opencv2/opencv.hpp>
+#include <stdexcept>
+#include <string>
+
+#include "layers/Layer.hpp"
+
+class InputLayer : public Layer {
+ public:
+  InputLayer() = default;
+  static void run(const std::vector<std::string>& path, Tensor& output) {
```
Comment (aobolensk) [inline include/layers/InputLayer.hpp:15]: run() should accept input tensor only. Image should be read separately outside any layer

File: include/layers/InputLayer.hpp [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+#include <iostream>
+#include <opencv2/opencv.hpp>
+#include <stdexcept>
+#include <string>
+
+#include "layers/Layer.hpp"
+
+class InputLayer : public Layer {
+ public:
+  InputLayer() = default;
```
Comment (aobolensk) [inline include/layers/InputLayer.hpp:14]: Input layer accepts parameters like:
- layout (NCHW, NHWC)
- coefficients for image normalization (mean, std, ...)

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: include/layers/InputLayer.hpp [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    std::vector<int> res(path.size() * 227 * 227 * 3);
+    for (int num = 0; num < static_cast<int>(path.size()); num++) {
+      cv::Mat image = cv::imread(path[num]);
+      if (image.empty()) {
+        throw std::runtime_error("Failed to load image");
+      }
+      cv::Mat resized_image;
+      cv::resize(image, resized_image, cv::Size(227, 227));
+      std::vector<cv::Mat> channels;
+      cv::split(resized_image, channels);
```
Comment (aobolensk) [inline include/layers/InputLayer.hpp:25]: This also should be outside of input layer. Image should be read and converted to Tensor to be processed by input layer

File: include/layers/InputLayer.hpp [hunk 12]
```diff
@@ -0,0 +1,41 @@
+#pragma once
+#include <algorithm>
+#include <cmath>
+#include <cstdlib>
+#include <iostream>
+#include <opencv2/opencv.hpp>
```
Comment (aobolensk) [inline include/layers/InputLayer.hpp:6]: Input layer should not depend of opencv, actually

File: include/layers/InputLayer.hpp (hunk 10 above)
Comment (AndreySorokin7) [reply include/layers/InputLayer.hpp:14]: And which one does we receive at the entrance of NCHW or NHWC? Or is it necessary to consider both options?

Comment (AndreySorokin7) [reply include/layers/InputLayer.hpp:20]: fixed

File: test/single_layer/test_inlayer.cpp [hunk 13]
```diff
@@ -0,0 +1,106 @@
+#include <cstdlib>
+#include <string>
+
+#include "gtest/gtest.h"
+#include "layers/InputLayer.hpp"
+
+using namespace itlab_2023;
+
+TEST(input, chech_basic) {
```
Comment (NeiroYT) [inline test/single_layer/test_inlayer.cpp:9]:
```suggestion
TEST(input, check_basic) {
```

**Review by NeiroYT — CHANGES_REQUESTED** [review]

File: include/layers/InputLayer.hpp [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+class InputLayer : public Layer {
+ private:
+  LayInOut layin_;
+  LayInOut layout_;
+  int mean_;
+  int std_;
+
+ public:
+  InputLayer() = default;
+  InputLayer(LayInOut layin, LayInOut layout, int mean, int std) {
```
Comment (NeiroYT) [inline include/layers/InputLayer.hpp:23]:
```suggestion
  InputLayer(LayInOut layin, LayInOut layout, int mean = 0, int std = 1) {
```
In case that we won't need normalization

**Review by aobolensk — APPROVED** [review]

File: 3rdparty/CMakeLists.txt [hunk 15]
```diff
@@ -1 +1,12 @@
...
+
+file(MAKE_DIRECTORY "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build")
+
+execute_process(
+    COMMAND ${CMAKE_COMMAND} -S "${CMAKE_SOURCE_DIR}/3rdparty/opencv" -B "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build" -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+    WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build"
+)
+execute_process(
+    COMMAND ${CMAKE_COMMAND} --build "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build" --config "${CMAKE_BUILD_TYPE}"
+    WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}/3rdparty/opencv/build"
```
Comment (aobolensk) [inline 3rdparty/CMakeLists.txt:11]: Follow up. Is that really needed

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: 3rdparty/CMakeLists.txt (hunk 15 above)
Comment (AndreySorokin7) [reply 3rdparty/CMakeLists.txt]: deleted

File: 3rdparty/CMakeLists.txt [hunk 16]
```diff
@@ -1 +1,2 @@
 add_subdirectory(googletest)
+
```
Comment (aobolensk) [inline 3rdparty/CMakeLists.txt:2]:
```suggestion
```

File: test/test_read/test_read.cpp [hunk 17]
```diff
@@ -9,6 +9,7 @@ TEST(Read_img, can_read_image) {
   const std::string image_path = std::string(TESTS_BINARY_PATH) + "/image.jpg";
   ASSERT_NO_THROW(Mat image = imread(image_path););
 }
+
```
Comment (aobolensk) [inline test/test_read/test_read.cpp:12]:
```suggestion
```

**Review by aobolensk — COMMENTED** [review]
Just a reminder, please avoid random newlines across the whole code base in the patches to keep them clean

File: app/CMakeLists.txt [hunk 18]
```diff
@@ -43,4 +43,3 @@ if (TENSORFLOW_FOUND)
 else()
     message(WARNING "TensorFlow not found. Skipping building of alexnet target.")
 endif()
```
Comment (aobolensk) [inline app/CMakeLists.txt:45]:
```suggestion
endif()

```

**Review by aobolensk — APPROVED** [review]

**Review by NeiroYT — APPROVED** [review]
